### DISCUSSION TITLE
What is the recommended way to create a model element (an AstNode) programatically?

### DISCUSSION THREAD

**Discussion by goto40** [eclipse-langium/langium#773] [Q&A] [answered]
What is the recommended way to create a model element (an AstNode) programatically?

I have a generator and I would like to create some artifical AstElements to pass to the generator.

* These objects shall be instances of an interface described by one grammar rule.
* These objects contain langium Reference<Type> attributes referencing other known model elements.

**Example:**
```
...
Definition: name=ID;
Property: definition=[Definition] value=INT;
...
```

In this example I would like to create a `Property` programatically and link it to a known `Definition`.

**Comment by spoenemann** [comment, marked as answer]
```typescript
const property: Property = {
    $type: Property,
    definition: {
        ref: definition,
        $refText: definition.name,
        $refNode: undefined!      // This is currently mandatory
    },
    value: 1234
};
// Add the new node to the AST
container.properties.push(property);
linkContentToContainer(container);
```

We should make `$refNode` optional to better support programmatic construction.

**Reply by spoenemann** [reply]
👉 https://github.com/langium/langium/pull/774
